### PR TITLE
RNNT Timestamps Fix & Timestamps in transcribe_speech_parallel

### PR DESCRIPTION
**Pull request by monica-sekoyan**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Fix the logic of counting tokens in offset by taking into consideration that the tokens for 1 frame are not always ending with blank (the cases when they can hit max_symbols and not add have blank token idx at the end)

Add timestamps extraction in `transcribe_speech_parallel.py`

+ other minor fixes

**Collection**: ASR

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to [#12795](https://github.com/NVIDIA/NeMo/issues/12795)

### CODE REVIEW

**Review by nithinraok — APPROVED** [review]
LGTM, can you add a unit test and also make sure this is handled for TDT

**Review by nithinraok — APPROVED** [review]
LGTM. Thanks!

File: nemo/collections/asr/models/rnnt_models.py [hunk 1]
```diff
@@ -815,7 +816,7 @@ def predict_step(self, batch, batch_idx, dataloader_idx=0):
         del signal
 
         best_hyp_text = self.decoding.rnnt_decoder_predictions_tensor(
-            encoder_output=encoded, encoded_lengths=encoded_len, return_hypotheses=False
+            encoder_output=encoded, encoded_lengths=encoded_len, return_hypotheses=True
```
Comment (nithinraok) [inline nemo/collections/asr/models/rnnt_models.py:819]: This should be handled for all models later gracefully as we return hypotheses always now.

File: nemo/collections/asr/parts/submodules/ctc_decoding.py [hunk 2]
```diff
@@ -792,10 +802,17 @@ def _get_word_offsets_chars(
                 end_offset = offset["end_offset"]
                 word += char
             else:
+                next_puntuation = (
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/ctc_decoding.py:805]: what was the issue here before?

**Review by nithinraok — APPROVED** [review]

File: nemo/collections/asr/parts/submodules/ctc_decoding.py (hunk 2 above)
Comment (monica-sekoyan) [reply nemo/collections/asr/parts/submodules/ctc_decoding.py:805]: there a was a space before punctuation marks with some tokenizers and this resulted only that punctuation mark being a separate word, like this 

```
{
      "word": "покупа",
      "start_offset": 22,
      "end_offset": 25
    },
    {
      "word": "поставите",
      "start_offset": 26,
      "end_offset": 29
    },
    {
      "word": ".",
      "start_offset": 31,
      "end_offset": 32
    }
```

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py [hunk 3]
```diff
@@ -1080,6 +1092,7 @@ def _after_inner_loop(self):
             self.state.labels,
             self.state.time_indices_current_labels,
             self.state.scores,
+            self.state.durations,
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py:1095]: Looks fine, but do we really need to store durations? 

Without durations - less to store, less complicated code.
If you need durations, you can pass the encoder tensor length to `batched_hyps_to_hypotheses` and get durations for free: 

```python
timestamps = batched_hyps.timestamps[i, : batched_hyps.current_lengths[i]]
durations =  torch.cat([timestamps[1:], encoder_lengths[i].unsqueeze(0)]) - timestamps
# use durations
...
```

So, my suggestion:
- remove durations from BatchedHyps and from the algorithm
- use timestamps to retrieve the durations

**Review by artbataev — APPROVED** [review]
LGTM. Let's refactor the decoding later to avoid costly alignments.
